src/lib/kernel/Linux.cc

```diff
@@ -141,30 +142,32 @@ int64_t Linux::faccessat(int64_t dfd, const std::string& filename, int64_t mode,
   // special file)
   new_pathname = Linux::getSpecialFile(filename);
 
-  // Get correct dirfd
-  int64_t dirfd = Linux::getDirFd(dfd, filename);
-  if (dirfd == -1) return EBADF;
+  // Get host dirfd. May return -1 in case of no mapping, pass through to host
+  // faccessat to deal with this
+  int64_t hostDfd = Linux::getHostDirFD(dfd);
 
   // Pass call through to host
-  int64_t retval = ::faccessat(dirfd, new_pathname.c_str(), mode, flag);
+  int64_t retval = ::faccessat(hostDfd, new_pathname.c_str(), mode, flag);
 
   return retval;
 }
 
-int64_t Linux::close(int64_t fd) {
+int64_t Linux::close(int64_t vfd) {
   // Don't close STDOUT or STDERR otherwise no SimEng output is given
   // afterwards. This includes final results given at the end of execution
-  if (fd != STDERR_FILENO && fd != STDOUT_FILENO) {
-    assert(fd < processStates_[0].fileDescriptorTable.size());
-    int64_t hfd = processStates_[0].fileDescriptorTable[fd];
+  if (vfd != STDERR_FILENO && vfd != STDOUT_FILENO) {
+    assert(vfd >= 0 && static_cast<size_t>(vfd) <
+                           processStates_[0].fileDescriptorTable.size());
+    int64_t hfd = processStates_[0].fileDescriptorTable[vfd];
     if (hfd < 0) {
+      // Early return, can't deallocate vfd that isn't in fileDescriptorTable
       return EBADF;
     }
 
     // Deallocate the virtual file descriptor
-    assert(processStates_[0].freeFileDescriptors.count(fd) == 0);
-    processStates_[0].freeFileDescriptors.insert(fd);
-    processStates_[0].fileDescriptorTable[fd] = -1;
+    assert(processStates_[0].freeFileDescriptors.count(vfd) == 0);
+    processStates_[0].freeFileDescriptors.insert(vfd);
```

Comment:
imilar to above, if we have changed the logic surrounding the asserts such that we now formally check the size of `freeFileDescriptors`, should this be done consistently?

Reply:
Applies throughout the rest of the file. I'm not bothered either way, but I think we should check if an index is within the bounds of a data structure in a consistent way

Reply:
See above RE directory fd vs fd